src/java.base/share/classes/java/io/BufferedInputStream.java

```diff
@@ -604,9 +605,15 @@ public long transferTo(OutputStream out) throws IOException {
     }
 
     private long implTransferTo(OutputStream out) throws IOException {
-        if (getClass() == BufferedInputStream.class
-                && ((count - pos) <= 0) && (markpos == -1)) {
-            return getInIfOpen().transferTo(out);
+        if (getClass() == BufferedInputStream.class && markpos == -1) {
+            int avail = count - pos;
+            if (avail > 0) {
+                // Prevent poisoning and leaking of buf
+                byte[] buffer = Arrays.copyOfRange(getBufIfOpen(), pos, count);
```

Comment:
@mkarg , could you please clarify why you added `Arrays.copyOfRange` here?
It seems to be an excessive copy that doesn't help much.

`buf` is `protected` in `BufferedInputStream`, so if someone really wants to get hold of the actual buffer, they can subclass `BufferedInputStream` and expose the buffer directly.
What do you think of removing `copyOfRange`?

Reply:
Buffer copy was not there before, and defensive copy was never present in `ByteArrayInputStream` as well: https://github.com/openjdk/jdk/blob/9a6ca233c7e91ffa2ce9451568b3be88ccd04504/src/java.base/share/classes/java/io/ByteArrayInputStream.java#L213

Reply:
Alan asked for this, and for good reason: While we implicitly trust subclasses as the buffer is "theirs" as part of the inheritance, we do not know where target stream comes from. It could be an injected verhicle to perform (at least) the following:
* Leaking data. The target stream could access data beyond the intended region.
* Poisoning. The target stream could write into the buffer.

Reply:
> As Alan pointed out, it is a bug (actually even a security risk), so BAIS should get fixed, too.

I am going to file an issue on this.

Reply:
> > As Alan pointed out, it is a bug (actually even a security risk), so BAIS should get fixed, too.
> 
> I am going to file an issue on this.

Thank you, I just planned to fix this today when I saw your existing PR! 👍